### PR TITLE
Ecm 97 - Member Days report

### PR DESCRIPTION
**Pull request by btensay**
A number of changes are added to implement the Member Days Report.

(https://tools.hmcts.net/jira/browse/ECM-97)

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ X ] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/memberdays/MemberDaysReport.java [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+import static java.util.stream.Collectors.groupingBy;
+import static uk.gov.hmcts.ecm.common.model.helper.Constants.HEARING_STATUS_HEARD;
+import static uk.gov.hmcts.ecm.common.model.helper.Constants.MEMBER_DAYS_REPORT;
+import static uk.gov.hmcts.ecm.common.model.helper.Constants.SINGLE_HEARING_DATE_TYPE;
+
+@Service
+@Slf4j
+public class MemberDaysReport {
+    private static final int MINUTES = 60;
+    private List<SubmitEvent> submitEvents;
```
Comment (jthmcts) [inline src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/memberdays/MemberDaysReport.java:30]: You can't use instance variables like these in Spring beans

**Review by jthmcts — CHANGES_REQUESTED** [review]

File: src/main/resources/application-dev.properties [hunk 2]
```diff
@@ -12,6 +12,12 @@ ccd.data.store.api.url=http://localhost:4452
 
 tornado.url=http://localhost:9095/api/render
 case_document_am.url =http://localhost:4455
-feature.secure-doc-store.enabled = ${SECURE_DOC_STORE_FEATURE:false}
+feature.secure-doc-store.enabled = true
 
-queue.create-updates.queue-name = ${CREATE_UPDATES_QUEUE_NAME}
\ No newline at end of file
+
+queue.create-updates.send.connection-string = Endpoint=sb://ethos-servicebus-devs.servicebus.windows.net/;SharedAccessKeyName=SendSharedAccessKey;SharedAccessKey=JpyFNVxQTESoFxCdwXtZXbz2MUT1Zlstdx7RLJL/0AA=;EntityPath=create-updates-tb
```
Comment (jthmcts) [inline src/main/resources/application-dev.properties:18]: This is your local dev settings. It shouldn't be in the commit. You don't need to change these values - use environment variables

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/service/TornadoService.java [hunk 3]
```diff
@@ -132,6 +133,7 @@ private void buildListingInstruction(HttpURLConnection conn, ListingData listing
             sb = ListingHelper.buildListingDocumentContent(listingData, tornadoConnection.getAccessKey(),
                     documentName, userDetails, caseType);
         }
+        log.info("In Tornado Service - Content of Document to generate: " + sb);
```
Comment (jthmcts) [inline src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/service/TornadoService.java:136]: What will this output to the logs?

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/interfaces/ReportDocPartial.java [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package uk.gov.hmcts.ethos.replacement.docmosis.reports.interfaces;
+
+import uk.gov.hmcts.ecm.common.model.listing.ListingData;
+
+public interface ReportDocPartial {
```
Comment (jthmcts) [inline src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/interfaces/ReportDocPartial.java:5]: Are you actually using this interface anywhere other than a class implementing it? What purpose does it serve?

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/memberdays/MemberDaysReport.java (hunk 1 above)
Comment (btensay) [reply src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/memberdays/MemberDaysReport.java:30]: corrected

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/service/TornadoService.java (hunk 3 above)
Comment (btensay) [reply src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/service/TornadoService.java:136]: It is the json representation of data that get fed to template. I only need that for dev. I take it out.

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/interfaces/ReportDocPartial.java (hunk 4 above)
Comment (btensay) [reply src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/interfaces/ReportDocPartial.java:5]: My thinking is that other report types should be refactored by implementing this interface. We are doing the template content building in one class now. At least, each report type can build & provide their specialised content. Right now I am using it in the current report. But I will use it for, say Claims Served, Members Sessions(which I work on next), etc.

Comment (jthmcts) [reply src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/interfaces/ReportDocPartial.java:5]: But you're not using the interface other than a class implements it. If that is the only usage then it provides no benefit. There's no point adding some partial code of a potential change

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/memberdays/MemberDaysReport.java (hunk 1 above)
Comment (jthmcts) [reply src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/memberdays/MemberDaysReport.java:30]: You can't use submitEvents and listingDetails as instance variables

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/memberdays/MemberDaysReport.java [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+    private void addReportDetails(MemberDaysReportData reportData) {
+        List<MemberDaysReportDetail> interimReportDetails = new ArrayList<>();
+        for (var submitEvent : submitEvents) {
+
+            if (!isValidCase(submitEvent)) {
+                continue;
+            }
+
+            var caseData = submitEvent.getCaseData();
+            log.info("Adding case {} to Member Days report", caseData.getEthosCaseReference());
```
Comment (jthmcts) [inline src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/memberdays/MemberDaysReport.java:77]: How many cases could you see on a report? Are we going to get logs full of large numbers of these messages?

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/memberdays/MemberDaysReport.java [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+                        reportDetail.setHearingDuration(getHearingDuration(listedItem));
+                        reportDetail.setParentHearingId(hearingItem.getId());
+                        interimReportDetails.add(reportDetail);
+                    }
+                }
+            }
+        }
+    }
+
+    private LocalDate getLocalDate(String listedDate) {
```
Comment (jthmcts) [inline src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/memberdays/MemberDaysReport.java:120]: Have you checked to see if there is already a helper method in ecm-common to do all this? You've got quite a lot of code to convert a string to a date - this is quite a common thing to do in ECM

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/memberdays/MemberDaysReportDoc.java [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+import static uk.gov.hmcts.ethos.replacement.docmosis.helpers.Helper.nullCheck;
+import static uk.gov.hmcts.ethos.replacement.docmosis.reports.Constants.REPORT_OFFICE;
+
+public class MemberDaysReportDoc {
+
+    public StringBuilder getReportDocPart(ListingData data) {
+        return getMemberDaysReport(data);
+    }
+
+    private static StringBuilder getMemberDaysReport(ListingData listingData) {
```
Comment (jthmcts) [inline src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/memberdays/MemberDaysReportDoc.java:19]: This doesn't need to be static

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/memberdays/MemberDaysReport.java (hunk 6 above)
Comment (hsjhita) [reply src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/memberdays/MemberDaysReport.java:120]: Or use SimpleDateFormatter which is already used in other classes, a similar implementation should be work where you can specify the exact pattern

File: src/main/resources/application-dev.properties [hunk 8]
```diff
@@ -12,6 +12,5 @@ ccd.data.store.api.url=http://localhost:4452
 
 tornado.url=http://localhost:9095/api/render
 case_document_am.url =http://localhost:4455
-feature.secure-doc-store.enabled = ${SECURE_DOC_STORE_FEATURE:false}
+feature.secure-doc-store.enabled = true
```
Comment (jthmcts) [inline src/main/resources/application-dev.properties:15]: You can just delete this line now as application.properties is 
feature.secure-doc-store.enabled = ${SECURE_DOC_STORE_FEATURE:true}

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/memberdays/MemberDaysReport.java (hunk 5 above)
Comment (btensay) [reply src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/memberdays/MemberDaysReport.java:77]: It is potentially(in theory) high as this section lists every single matching hearing date, that could be multiple even in one case. Hence, I have taken out the log message.

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/memberdays/MemberDaysReport.java (hunk 6 above)
Comment (btensay) [reply src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/memberdays/MemberDaysReport.java:120]: I am having look.

**Pull request closed without merging**